ibis/common/caching.py

```diff
@@ -47,56 +49,50 @@ def __init__(
         generate_name: Callable[[], str],
         key: Callable[[Any], Any],
     ) -> None:
-        self.cache = bidict()
-        # Somehow mypy needs a type hint here
-        self.refs: Counter = Counter()
         self.populate = populate
         self.lookup = lookup
         self.finalize = finalize
-        # Somehow mypy needs a type hint here
-        self.names: defaultdict = defaultdict(generate_name)
+        self.generate_name = generate_name
         self.key = key or (lambda x: x)
 
+        self.cache: dict[Any, CacheEntry] = dict()
+
     def get(self, key, default=None):
-        try:
-            return self[key]
-        except KeyError:
-            return default
+        if (res := self.cache.get(key)) is not None:
+            return res.ref()
+        return default
 
     def __getitem__(self, key):
-        result = self.cache[key]
-        self.refs[key] += 1
-        return result
+        return self.cache[key].ref()
```

Comment:
Should this raise a `KeyError` if it returns `None`?

Reply:
It needs to loose a race for that to happen (since the reference invalidation triggers the removal of the cache entry) and `__getitem__` is unused, but you are right!